lib/srv/uacc/uacc_linux.go

```diff
@@ -234,6 +244,61 @@ func UserWithPtyInDatabase(utmpPath string, username string) error {
 	}
 }
 
+// LogFailedLogin writes a new entry to the btmp failed login log.
+// This should be called when an interactive session fails due to a missing
+// local user.
+//
+// `username`: Name of the user the interactive session is running under.
+// `hostname`: Name of the system the user is logged into.
+// `remote`: IPv6 address of the remote host.
+func LogFailedLogin(btmpPath, username, hostname string, remote [4]int32) error {
+	// String parameter validation.
+	if len(username) > userMaxLen {
+		return trace.BadParameter("username length exceeds OS limits")
+	}
+	if len(hostname) > hostMaxLen {
+		return trace.BadParameter("hostname length exceeds OS limits")
+	}
+
+	btmpPath = getDefaultBtmpPath(btmpPath)
+	// Convert Go strings into C strings that we can pass over ffi.
+	cBtmpPath := C.CString(btmpPath)
+	defer C.free(unsafe.Pointer(cBtmpPath))
+	cUsername := C.CString(username)
+	defer C.free(unsafe.Pointer(cUsername))
+	cHostname := C.CString(hostname)
+	defer C.free(unsafe.Pointer(cHostname))
+
+	// Convert IPv6 array into C integer format.
+	cIP := [4]C.int{0, 0, 0, 0}
+	for i := 0; i < 4; i++ {
+		cIP[i] = (C.int)(remote[i])
```

Comment:
In C the parameter is `const int32_t remote_addr_v6[4]`, so I think we should use `C.int32_t` here.
```suggestion
		cIP[i] = (C.int32_t)(remote[i])
```

Seeing as the original code in `Open` has the same issue, it might be beneficial to refactor this code to use shared helper functions.

Reply:
Came to ask the same thing. You're memcpying `4 * sizeof(int32_t)`, but on 64-bit platforms this array is probably twice as big.

Also, arrays are zero-valued by default in Go. `cIP := [4]C.int{0, 0, 0, 0}` is equivalent to `var cIP [4]C.int`.